### DISCUSSION TITLE
Internal Comment Engine

### DISCUSSION THREAD

**Discussion by jsalgado10** [OrchardCMS/OrchardCore#7552] [Ideas]
Aside from using Disqus, Is there another way to add comments to a content type and maintain those on the same database as Orchard Core?

**Comment by hishamco** [comment]
Nothing prevent you from doing that, AFAIK Orchard 1 Blogs has comments, you can create a part that could be attached to blog post

**Comment by sebastienros** [comment]
Tracking issue:
https://github.com/OrchardCMS/OrchardCore/issues/2942

Someone who is looking into it:
https://github.com/StatCan/StatCan.OrchardCore/issues/73

**Reply by jsalgado10** [reply]
I will start my own module that will add a comments engine. To avoid having to create a new database table, two new content types will be needed and an Api controller to handle CRUD operations. 

I think this will be enough so anybody can use the module with their own custom views and just have to call the api methods to retrieve the comments for each article. 

I am worried about performance of the site when having a lot of comments. Will a new database table work much better?

Any suggestions are more than welcome

**Reply by sebastienros** [reply]
You can try to look at the suggestions I made in the issue I referenced. I am talking about how paging can be done.
Comments are hard. Mostly due to how people like their comments: threaded or not, ordered by what, ranking or not, markdown or text, spam, perf, anonymous access or not, ...

And for all these issues there are no single answer. Maybe take a look at what other platforms do, like Drupal, CraftCMS, Diqus, ...
If you intend to work on a PR for this repos, which we would welcome, feel free to ping me for help on design decisions.

**Reply by jsalgado10** [reply]
I have looked at Disqus and they are using a embed.js file to draw the comment layout inside an iframe. Not a bad idea but i would like a more customizable way to display the comments like you suggested.  

Yes I would like to work on a PR, but first i will try to finish what i currently have and have a prototype to present

**Reply by jsalgado10** [reply]
@sebastienros I am currently working on the controller that will handle all the CRUD operations for the comments

I created a new content type CommentPost that has all the details for each comment. I will be setting the DocumentId that it belongs to and the CommentParentId if this was a reply to another comment. 

I am getting all records of ContentType "CommentPost"
List<ContentItem> contentItems;
            contentItems = (await session.Query<ContentItem, ContentItemIndex>()
                .Where(x => x.ContentType == "CommentPost" && x.Latest == true )
                .ListAsync()).ToList();

But I want to filter out all records where documentid = currentdocumentid sent from request.

To Display each comment record I will be using the BuildDisplayAsync method. If understand correctly, this will provide a way to override the display template with a custom liquid template.
var contentItemSummaries = new List<dynamic>();
            foreach (var contentItem in contentItems)
            {
                contentItemSummaries.Add(await _contentItemDisplayManager.BuildDisplayAsync(contentItem, _updateModelAccessor.ModelUpdater, "Summary"));
            }

The main question is how to filter out the records on the list by documentid

**Reply by jsalgado10** [reply]
@sebastienros so finally I completed the module and I am working on adding a Pager. However, since the controller is returning a partial view and I am updating the div using jquery, I do not want to take the user to a different page as it will not show with the correct theme formatting. The question is if I can override the Pager shape and call a javascript function when the user select the page number

**Comment by jsalgado10** [comment]
Here is what I have so far. 

https://github.com/jsalgado10/OrchardCore.Commentator
